### PR TITLE
feature: add query as object constructor and tests

### LINKED ISSUE #366: URI constructor does not take an object as value for query property
Hello everyone,

Please let me know if this is this has been decided to be like this or not, but when building an URL, query property only takes a string as value, instead of search() or query() that take both string and objects. 

Heres an example code ([online repl here](https://repl.it/repls/AshamedRoundPangolin)):
```js 
const assert = require('assert');
const URI = require('urijs');

const PROTOCOL = 'http';
const HOST = 'mycustom.domain.com';
const PORT = '443';
const BASE_PATH = '/some//path/';
const resource = '///resource//5';
const query = {
  myQuery: 'rocks'
}

const constructorURI = new URI({
    protocol: PROTOCOL,
    hostname: HOST,
    port: PORT,
    path: `${BASE_PATH}/${resource}/`,
    query,
  }).normalize();
  
  
const searchURI = new URI({
    protocol: PROTOCOL,
    hostname: HOST,
    port: PORT,
    path: `${BASE_PATH}/${resource}/`,
  }).search(query).normalize();

assert(constructorURI.toString() === searchURI.toString(), 'URIs are not equal');
```

### PR DESCRIPTION
**Pull request by wachunei**
Attempt to resolve #366

### CODE REVIEW

File: src/URI.js [hunk 1]
```diff
@@ -1241,10 +1241,14 @@
...
       var src = _URI ? href._parts : href;
       for (key in src) {
+        if (key === 'query') {
+          continue;
+        }
         if (hasOwn.call(this._parts, key)) {
           this._parts[key] = src[key];
         }
       }
+      this.query(src['query']);
```
Comment (rodneyrehm) [inline src/URI.js]: this should probably only be executed if `src.query` is actually defined

Comment (wachunei) [reply src/URI.js]: definitely! will update.

File: src/URI.js [hunk 2]
```diff
@@ -1240,15 +1240,19 @@
       this._parts = URI.parse(String(href), this._parts);
     } else if (_URI || _object) {
       var src = _URI ? href._parts : href;
+      var hasQueryPart = false;
       for (key in src) {
         if (key === 'query') {
+          hasQueryPart = true;
```
Comment (wachunei) [inline src/URI.js]: wait a second, this can be true even if `src['query']` is undefined, isnt it?

Comment (rodneyrehm) [reply src/URI.js]: you can keep the `if (key === 'query') { continue; }` part and add `if (src.query) { this.query(src.query, false); }`. you only need to execute .query() if there actually is a value (other than null, undefined, empty-string)

**Comment by wachunei** [comment]
About fragments: I think we should open its own issue and solve it in its own PR.

File: test/test.js [hunk 3]
```diff
@@ -34,6 +34,49 @@
...
+  });
+  test('new URI(object)', function() {
+    var u = new URI({
+      protocol: 'http',
+      hostname: 'example.org',
+      query: 'foo=bar&bar=foo',
+    });
+    ok(u instanceof URI, 'instanceof URI');
+    ok(typeof u.query() === 'string', 'query is string');
+    ok(u.query() === 'foo=bar&bar=foo', 'query has right value');
```
Comment (rodneyrehm) [inline test/test.js]:
```js
equal(u.query(), 'foo=bar&bar=foo', 'query() value');
```

File: test/test.js [hunk 4]
```diff
@@ -34,6 +34,49 @@
...
+    var u = new URI({
+      protocol: 'http',
+      hostname: 'example.org',
+      query: '?foo=bar&bar=foo',
+    });
+    ok(u instanceof URI, 'instanceof URI');
+    ok(typeof u.query() === 'string', 'query is string');
+    ok(u.query() === 'foo=bar&bar=foo', 'query has right value');
+    ok(u.search() === '?foo=bar&bar=foo', 'search has right value');
+    ok(typeof u.query(true) === 'object', 'query map is object');
```
Comment (rodneyrehm) [inline test/test.js]:
```js
deepEqual(u.query(true), { foo: 'bar', bar: 'foo' }, 'query(true) value');
```

**Comment by rodneyrehm** [comment]
> About fragments: I think we should open its own issue and solve it in its own PR.

fine by me :)

**Comment by rodneyrehm** [comment]
thank you!

**Comment by wachunei** [comment]
You're welcome, looking forward to this being released! 👍

**Pull request merged**